common/lib/dependabot/file_parsers/base/dependency_set.rb

```diff
@@ -50,41 +60,101 @@ def case_sensitive?
           @case_sensitive
         end
 
-        def dependency_for_name(name)
-          return dependencies.find { |d| d.name == name } if case_sensitive?
+        def key_for_name(name)
+          case_sensitive? ? name : name.downcase
+        end
 
-          dependencies.find { |d| d.name&.downcase == name&.downcase }
+        def key_for(dep)
+          key_for_name(dep.name)
         end
 
-        def combined_dependency(old_dep, new_dep)
-          package_manager = old_dep.package_manager
-          v_cls = Utils.version_class_for_package_manager(package_manager)
-
-          # If we already have a requirement use the existing version
-          # (if present). Otherwise, use whatever the lowest version is
-          new_version =
-            if old_dep.requirements.any? then old_dep.version || new_dep.version
-            elsif !v_cls.correct?(new_dep.version) then old_dep.version
-            elsif !v_cls.correct?(old_dep.version) then new_dep.version
-            elsif v_cls.new(new_dep.version) > v_cls.new(old_dep.version)
-              old_dep.version
+        class DependencySlot
+          attr_reader :combined
+
+          def initialize
+            @all_versions = []
+            @combined = nil
+          end
+
+          def all_versions(sort:)
+            sort ? sorted : @all_versions.dup
```

Comment:
Why do we need to dup all_versions only if we're not sorting? 🤔

Reply:
Sorting produces a new array, so I wanted to mirror that behavior and make a shallow copy. Helps reduce the risk of a caller will accidentally mutating `@all_versions`.

Reply:
I removed the sort option and made `all_versions` an attr reader so this is moot now.